synapse/storage/databases/main/search.py

```diff
@@ -372,6 +377,28 @@ def reindex_search_txn(txn):
 
         return num_rows
 
+    async def _background_delete_non_strings(
+        self, progress: JsonDict, batch_size: int
+    ) -> int:
+        """Deletes rows with non-string `value`s from `event_search` if using sqlite.
+
+        Prior to Synapse 1.44.0, malformed events received over federation could cause integers
+        to be inserted into the `event_search` table when using sqlite.
+        """
+
+        def delete_non_strings_txn(txn: LoggingTransaction) -> None:
+            txn.execute("DELETE FROM event_search WHERE typeof(value) != 'text'")
```

Comment:
Do we need to limit this in any way? (Will this cause any locks on the table?)

Reply:
Good question! It crossed my mind at some point and I forgot to make a note of it on the PR.

`event_search` is defined as a full text search virtual table with no explicit indexes:
`CREATE VIRTUAL TABLE event_search USING fts4 (event_id, room_id, sender, key, value);`
so it looked like doing the processing in batches would be slower, since DELETEing by specific event_ids would have to scan the whole table anyway.

I think this will lock the table for the duration of the DELETE. The sqlite docs suggest that locks are done on the file level, so it's actually the entire database that gets locked, which is unfortunate.

We could add an index first, then delete in batches, but if we're going to ask sqlite to scan the whole table to build an index, we might as well just do the delete.